wip/value-domain.md

```diff
@@ -0,0 +1,128 @@
+# Rust Value Domain
+
+**Note:** This document is not normative nor endorsed by the UCG WG. Its purpose is to be the basis for discussion and to set down some key terminology.
+
+The purpose of this document is to describe what the set of *all possible values* is in Rust.
+This is an important definition: one key element of a Rust specification will be to define the [representation relation][representation] of every type.
+This relation relates values with lists of bytes: it says, for a given value and list of bytes, if that value is represented by that list.
+However, before we can even start specifying the relation, we have to specify the involved domains.
+`Byte` is defined as part of [the memory interface][memory-interface]; this document is about defining `Value`.
+
+[representation]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/reference/src/glossary.md#representation
+[memory-interface]: memory-interface.md
+
+## Values
+
+The Rust value domain is described by the following (incomplete) type definition (in terms of an arbitrary `Pointer` type instantiated by [the memory interface][memory-interface]):
+
+```rust
+enum Value<Pointer> {
+    /// A mathematical integer.
+    Int(BigInt),
+    /// A Boolean value.
+    Bool(bool),
+    /// A pointer.
+    Ptr(Pointer),
+    /// An uninitialized value.
+    Uninit,
+    /// An n-tuple.
+    Tuple(Vec<Self>),
+    /// A variant of a sum type.
+    Variant {
+        idx: u64,
+        data: Box<Self>,
+    },
+    /// A "bag of raw bytes".
+    RawBag(Vec<Byte<Pointer>>),
```

Comment:
Why is this needed ? An array should be able to cover this.

Reply:
I don't follow. No other variant of `Value` even lets you have a `Byte`.

Reply:
Oh, I thought that BigInt would let you have a byte. So if aI cast a pointer to an `usize`, is that usize a `BigInt` or a `RawBag` ?

Reply:
I meant `Byte` as in what the memory-interface defines. `BigInt` is a mathematical integer, as the comment indicates. So when you cast a ptr to an int, you get a mathematical integer and that is represented via the `Int` variant.

`RawBag` is for unions.

Reply:
I added a note saying `RawBag` is for unions, and hope this answers your question.